prog/vm/github_runner.rb

```diff
@@ -260,6 +260,27 @@ def before_run
 
     if vm
       vm.private_subnets.each { _1.incr_destroy }
+
+      # If the runner is not assigned any job and we destroy it after a
+      # timeline, the workflow_job is nil, in that case, we want to be able to
+      # see journalctl output to debug if there was any problem with the runner
+      # script.
+      #
+      # We also want to see the journalctl output if the runner script failed.
+      #
+      # Hence, the condition is added to check if the workflow_job is nil or
+      # the conclusion is failure.
+      if !github_runner.workflow_job || github_runner.workflow_job.fetch("conclusion", "failure") == "failure"
+        begin
+          serial_log_path = "/vm/#{vm.inhost_name}/serial.log"
+          vm.vm_host.sshable.cmd("sudo mv -f #{serial_log_path} /var/log/ubicloud/serials/#{github_runner.ubid}_serial.log")
```

Comment:
We haven't destroyed the virtual machine at this point. CH is still running. I believe we shouldn't move the serial log file. Using `cp` might be a better option. @fdr knows better

Reply:
We will, though. I don't see an issue here.

Reply:
The outcome here will depend. If mv reduces to the "rename" system call because the target is on the same file system, then cloud hypervisor will continue to write to the same file presuming it does the simple thing  and leaves the file descriptor open. That means you will get all the residual output to the very end.

If it's on another file system, then mv becomes cp+rm. in this case, the open file descriptor will refer to an unlinked file, so the file system will not reclaim space until the cloud hypervisor closes the descriptor. The copy, for which cloud-hypervisor does not have a file descriptor, will not get any residual updates.

I think that you should consider switching to a hard link (ln without -s). This will not work between file systems and has the same file appear twice in the hierarchy (via two directory entries to the same file), this means that any uses of serial.log in the /vm location continue to work in an oblivious manner. 

All names must be unlinked (deleted) and have their file descriptors closed to reclaim space.

Reply:
Oh, one thing to watch out for is uid recycling when the vm user is deleted. It would be fastidious to intrude before user deletion to rewrite ownership of that file.

Reply:
We are actually moving the file in between file systems. It was under `/vm/vmxxxx/serial.log` and we move it to `/var/log/ubicloud/serials/xxxx_serial.log`. I don't think the problem here that enes is referring to is getting the residual updates but any chance of creating a failure to the destroy code path. Or, if something goes wrong at the destroy time and we want to debug it by looking into serial.log but "hey! it's not there?!?". 
Considering all, I agree `cp` is a better option. 
@fdr does this sound good to you?

Reply:
It's sort of dissatisfying to copy bytes like that unless we really need to, but serial.log is small enough and our needs dire enough. How about using "cp" then?

Reply:
… also are you sure those are different file systems? Did we mount /vm specially?

Reply:
You're right, I was wrong. They are in the same filesystem. I have updated it to use `ln`